blazor-application/Pages/BLEPage.razor

```diff
@@ -0,0 +1,167 @@
+    </div>
+    <br />
+    <p>N = @(n)</p>
+    <div style="display: flex; flex-direction: row; align-items: center;">
+        <input type="range" min="2" max="4" step="0.01" @bind="n" @bind:event="oninput" style="width: 100%" />
+    </div>
+    <br />
+    <p>Estimated distance: @(Math.Round(LogDistancePathLoss(connectedToDevice.Rssi), 2)) m</p>
+    <p>Average distance: @(Math.Round(AverageDistance(), 2)) m</p>
+    <button @onclick=ClearMeasurements>Clear measurements</button>
+}
+else 
+{
+    <table class="table">
+        <thead>
+            <tr>
+                <th>Name</th>
+                <th>Id</th>
+                <th></th>
+            </tr>
+        </thead>
+        <tbody>
+            @foreach (IDevice device in BLEService.Devices)
+            {
+                <tr @onclick="() => OnClickConnect(device)">
+                    <td>@(GetDeviceDisplayName(device))</td>
+                    <td>@device.Id</td>
+                </tr>
+            }
+        </tbody>
+    </table>
+}
+
+@code {
+    [Inject]
+    private IPromptService promptService { get; set; }
+
+    private bool isConnecting;
+
+    private double txPower = -50;
+    private double n = 3;
+
+    private List<double> measuredDistances = new List<double>();
+
+    private PeriodicTimer periodicTimer;
+
+    private IDevice connectedToDevice;
+
+    protected override void OnInitialized()
+    {
+        base.OnInitialized();
+        BLEService.DevicesChanged += StateHasChanged;
+    }
+
+    protected override async Task OnInitializedAsync()
+    {
+        await InvokeAsync(BLEService.BeginScan);
+    }
+
+    private void OnClickConnect(IDevice device)
+    {
+        promptService.ShowConfirmation("Connect to device", $"Would you like to connect to the device \"{GetDeviceDisplayName(device)}\"?", (accepted) =>
+        {
+            if (accepted)
+            {
+                isConnecting = true;
+                InvokeAsync(StateHasChanged);
+
+                BLEService.ConnectToDevice(device, (success, msg) =>
+                {
+                    if (success)
+                    {
+                        connectedToDevice = device;
+                        connectedToDevice.UpdateConnectionInterval(ConnectionInterval.High);
+                        promptService.ShowAlert("Connection succeeded", $"Successfully connected to the device \"{GetDeviceDisplayName(device)}\"!");
+                        InvokeAsync(UpdateRssiPeriodically);
+                    }
+                    else
+                    {
+                        promptService.ShowAlert("Connection failed", $"Could not connect to device: {msg}.");
+                    }
+
+                    isConnecting = false;
+                    InvokeAsync(StateHasChanged);
+                });
+            }
+        });
+    }
+
+    private async void UpdateRssiPeriodically()
```

Comment:
We might want to consider wrapping this function in some type periodically executing some action? Maybe `PeriodicExecutor` It seems that we are likely to perform multiple operations for 'scanning periodically' during this project.

Reply:
Maybe later, but I do not like speculative engineering and therefore see no reason to do this now when we currently have no use for it.